### PR TITLE
[WIP] New Homepage for ESPnet

### PR DESCRIPTION
**Pull request by Masao-Someki**
The new homepage proposal with Sphinx and VuePress! 
You can see the demo-page from the following link: https://masao-someki.github.io/espnet_draft_home_page/

| Landing page | API Documents | Notebooks | Search box |
| --- | --- | --- | --- |
|  image  |  image |   image | image |

## What?

The `./ci/doc.sh` script will automatically generate the HTML files for the documentation. Performance has been optimized, as evidenced by the Lighthouse results.

| from PC | from mobile |
| --- | --- |
| image | image |

## Why?

The current documentation is difficult to integrate with the Notebook repository because Sphinx generates docstrings in HTML format, while the notebooks are in Jupyter Notebook format. Additionally, changing the directory structure is challenging, leading to inefficient maintenance.

To address this, we convert all documents into Markdown format and generate the HTML files using VuePress. With VuePress, we can define the page structure based on the directory structure of each Markdown file, eliminating the need to create a TOC or specific reStructuredText (RST) files. This approach also allows us to easily switch to other static site generators, such as [MkDocs](https://www.mkdocs.org/) and [Hugo](https://gohugo.io/).

## ToDO
- [x] Fix bug in script
- [x] Add link or button to go to source of the package

### CODE REVIEW

**Comment by juice500ml** [comment]
Currently struggling while building this. It seems that the current version of sphinx is severely outdated, and we should consider upgrading it. For example, recommonmark is now deprecated. https://github.com/readthedocs/recommonmark/issues/221

**Comment by sw005320** [comment]
> Currently struggling while building this. It seems that the current version of sphinx is severely outdated, and we should consider upgrading it. For example, recommonmark is now deprecated. [readthedocs/recommonmark#221](https://github.com/readthedocs/recommonmark/issues/221)

It sounds good to me to update Sphinx.
But Sphinx usually has complicated dependencies, so we fixed the version.
So, I hope it will not bring further complications.

**Comment by Masao-Someki** [comment]
I apologize for the delayed response, could you please specify where the problem occurs? Is it related to Sphinx?

> It seems that the current version of sphinx is severely outdated, and we should consider upgrading it. For example, recommonmark is now deprecated. https://github.com/readthedocs/recommonmark/issues/221

Thank you, I will look into it.

**Comment by juice500ml** [comment]
Yes, I'm a bit worried about the implications. I'd usually avoid updating too much... If you use Python 3.10, I think that problem persists. Let me try downgrading Python and run it again. Actually, nearly all the steps for espnet/ci/doc.sh raises issues for me. Possibly because I only installed `espnet[doc]` via pypi on a barebone Python?

**Comment by Masao-Someki** [comment]
Would you reinstall ESPnet with all the necessary dependencies, such as `espnet[doc]` or `espnet[recipe]`, and run the command again? If you encounter any issues with the updated ESPnet, please let me know…

**Comment by sw005320** [comment]
Can you fix the CI errors?

**Comment by Masao-Someki** [comment]
Thank you, @juice500ml! The CI for document generation is now working successfully!

**Comment by sw005320** [comment]
We still have several CI errors.
Please fix them

**Comment by sw005320** [comment]
Did you try it in your local environment?
It would be easier to fix the error than to ask CI often.

**Comment by Masao-Someki** [comment]
I'm really sorry about this, but when I tested it in my local environment, I encountered another error. It seems like it might take a while to resolve, so I'm running the jobs now...

**Comment by sw005320** [comment]
I see. Make sense.
But since this is a CPU job, we can build a virtual Linux environment in your terminal PC and create exactly the same setup by following https://github.com/espnet/espnet/tree/master/.github/workflows

**Comment by Masao-Someki** [comment]
@juice500ml

It looks like the `gtn` package wasn't installed during the `install espnet` process in the CI log from #5766. The `gtn` seems to be included in the current CI because the `all` dependency group is added to the `doc` group:

https://github.com/Masao-Someki/espnet/blob/386b291df6e4db4272bddddb19a4ca4877186950/setup.py#L123

```
requirements["doc"].extend(requirements["all"])
```

I think we can resolve the CI issues by removing this. Could you please confirm if it's okay to revert this line?

**Comment by juice500ml** [comment]
I'm pretty sure that CI will work now. I'm planning to make another PR (within today) that makes updates on the doc design, etc. based on the feedbacks from other people. I don't want to make the PR to be bigger than this :)

File: doc/usage2rst.sh [hunk 1]
```diff
@@ -8,6 +8,8 @@ if [ $1 == "--help" ]; then
 fi
 
 real=$(realpath $1)
+# githash=`git rev-parse HEAD`
+githash=8574aa8e0d5968916df1df42c28621a48b2a5281
```
Comment (sw005320) [inline doc/usage2rst.sh:12]: what's this?
is it OK to fix the hash?

**Review by sw005320 — COMMENTED** [review]
- Thanks. I agree that the PR should not be bigger than this. I'll just wait for the next PR to check the content, but anyway, can @Masao-Someki and @juice500ml to cross-review this PR again to make sure?
- I'm just concerned that `setup.py` specifies the version or restricts the version. If they are necessary, it's OK

File: doc/usage2rst.sh (hunk 1 above)
Comment (Masao-Someki) [reply doc/usage2rst.sh:12]: @sw005320 
Sorry, this is a debug code! I will check all diffs and fix debug lines!!

**Comment by Masao-Someki** [comment]
Regarding the modifications to `setup.py`:

- We restricted `setuptools` to `<74.0.0` because the latest version (`74.0.0`) removed a functionality related to `distutils`, which would cause an installation error.
- We upgraded Sphinx to the latest version, adding a version restriction to prevent potential CI errors in case of a major Sphinx update.
- Due to the Sphinx upgrade, `recommonmark` was deprecated. The recommended replacement for Markdown support is `myst-parser`.
- We've added a Jupyter notebook as a demo. To convert `.ipynb` files to Markdown, we now use `jupyterlab`.
- We are using `sphinx-markdown-builder` to build Markdown files from RST files.

**Comment by juice500ml** [comment]
I reviewed the PR & refactored some minor stuff. I think it's okay now, so let's wait for the checks to finish!

**Comment by sw005320** [comment]
OK
After CI is fixed, I'll merge this PR.

**Pull request merged**